folding/src/expressions.rs

```diff
@@ -24,6 +24,186 @@
+//! degree `d` by introducing the new variable `u`, and introduce the concept of
+//! "error terms" that will englobe the "cross-terms". The prover takes care of
+//! computing the cross-terms and commit to it.
+//!
+//! While folding, we aggregate the error terms of all instances into a single
+//! error term, E.
+//! For instance, with our example, if we have a folded instance with the non-zero
+//! error terms `E_{1}` and `E_{2}`, we have:
+//! ```text
+//! E = E_{1} + r T + E_{2}
+//! ```
+//!
+//! ## Aggregating constraints
+//!
+//! The library also provides a way to fold NP relations described by a list of
+//! multi-variate polynomials, like we usually have in a zkSNARK circuit.
+//!
+//! In PlonK, we aggregate all the polynomials into a single polynomial by
+//! coining a random value `α`. For instance, if we have two polynomials `P` and
+//! `Q` describing our computation in a zkSNARK circuit, we usually use the
+//! randomized polynomial `P + α Q` (used to build the quotient polynomial in
+//! PlonK).
+//!
+//! More generally, if for each row, our computation is constrained by the polynomial
+//! list `[P_{1}, P_{2}, ..., P_{n}]`, we can aggregate them into a single
+//! polynomial `P_{agg} = ∑_{i} α^{i} P_{i}`.
+//!
+//! In particular, when we reduce a polynomial to degree 2, we have this case
+//! where the circuit is described by a list of polynomials and we aggregate
+//! them into a single polynomial.
+//!
+//! For instance, if we have two polynomials `P(X_{1}, X_{2}, X_{3})` and
+//! `Q(X_{1}, X_{2}, X_{3})` such that:
+//! ```text
+//! P(X_{1}, X_{2}, X_{3}) = X_{1} * X_{2} + (1 - X_{3})
+//! Q(X_{1}, X_{2}, X_{3}) = X_{1} + X_{2}
+//! ```
+//!
+//! The relaxed form of the polynomials are:
+//! ```text
+//! P_relaxed(X_{1}, X_{2}, X_{3}, u) = X_{1} * X_{2} + u (u - X_{3})
+//! Q_relaxed(X_{1}, X_{2}, X_{3}, u) = u X_{1} + u X_{2}
+//! ```
+//!
+//! We start by coining `α_{1}` and `α_{2}` and we compute the polynomial
+//! `P'(X_{1}, X_{2}, X_{3}, u, α_{1})` and `Q'(X_{1}, X_{2}, X_{3}, α_{2})` such that:
+//! ```text
+//! P'(X_{1}, X_{2}, X_{3}, u, α_{1}) = α_{1} P_relaxed(X_{1}, X_{2}, X_{3}, u)
+//!                                   = α_{1} (X_{1} * X_{2} + u (u - X_{3}))
+//!                                   = α_{1} X_{1} * X_{2} + α_{1} u^2 - α_{1} u X_{3}
+//! Q'(X_{1}, X_{2}, X_{3}, u, α_{2}) = α_{2} Q_relaxed(X_{1}, X_{2}, X_{3}, u)
+//!                                   = α_{2} (u X_{1} + u X_{2})
+//!                                   = α_{2} u X_{1} + α_{2} u X_{2}
+//! ```
+//! and we want to fold the multivariate polynomial S defined over six
+//! variables:
+//! ```text
+//!   S(X_{1}, X_{2}, X_{3}, u, α_{1}, α_{2})
+//! = P'(X_{1}, X_{2}, X_{3}, u, α_{1}) + Q'(X_{1}, X_{2}, X_{3}, u, α_{2})`.
+//! = α_{1} X_{1} X_{2} +
+//!   α_{1} u^2 -
+//!   α_{1} u X_{3} +
+//!   α_{2} u X_{1} +
+//!   α_{2} u X_{2}
+//! ```
+//!
+//! Note that we end up with everything of the same degree, which is `3` in this
```

Comment:
Maybe here it should be clarified that the powers of alpha are considered as degree 1 so the degree does not grow further than 3.

Reply:
What do you think of e8674cdc2c01451a7849bc7192f9e29c063c802e?